### DISCUSSION TITLE
BIND_TABLE some new features

### DISCUSSION THREAD

**Discussion by RixarSAP** [abap2xlsx/abap2xlsx#1223] [Ideas]
Hello, all.

I've been using ABAP2XLSX for the last to weeks, so I still need to learn a lot how to use it properly.
But creating some reports I miss some features that can be useful.

I've made some changes in the following methods, I let you know and in case you like the idea I can create a new pull request with these changes.

I've added these two new fields in field catalog structure

![image](https://github.com/abap2xlsx/abap2xlsx/assets/168567887/06eb2c88-add8-4238-99b8-89512885b159)

The first field COLUMN_LENGHT is to change the width of the columns with the lenght set in the field catalog. And avoid call SET_COLUMN_WIDTH several times after "print" the table in the worksheet.

 I've changed the method BIND_TABLE adding these lines:

```
*   It is better to loop column by column (only visible column)
    LOOP AT lt_field_catalog ASSIGNING <ls_field_catalog> WHERE dynpfld EQ abap_true.

      lv_column_alpha = zcl_excel_common=>convert_column2alpha( lv_column_int ).

*    Lines added:
      IF <ls_field_catalog>-column_lenght IS NOT INITIAL.
        set_column_width( ip_column = lv_column_alpha ip_width_fix = <ls_field_catalog>-column_lenght ).
      ENDIF.
```

The second field, EDIT_MASK is to add the possibility to use an edit_mask, for this field I've also changed the method BIND_TABLE with this code (pending review if it can be done in a better way)

You can fill manually this field in your fieldcatalog with values like '==ALPHA' to remove the leading zeros.

```
        lv_value = <fs_fldval>.
        IF <ls_field_catalog>-edit_mask IS NOT INITIAL. "Rircardo R.
          TRY.
*            WRITE <fs_fldval> TO lv_value USING EDIT MASK <ls_field_catalog>-edit_mask. "Cannot used with a String
              lv_mask_fm = <ls_field_catalog>-edit_mask.
              SHIFT lv_mask_fm LEFT DELETING LEADING '='. "Remove first two =
              CONCATENATE 'CONVERSION_EXIT_' lv_mask_fm '_OUTPUT' INTO lv_mask_fm.
              CALL FUNCTION lv_mask_fm
                EXPORTING
                  input  = <fs_fldval>
                IMPORTING
                  output = lv_value.
            CATCH cx_root.
              lv_errormessage = 'Error converting field value using MASK &1'.
              REPLACE '&1' WITH <ls_field_catalog>-edit_mask INTO lv_errormessage.
              zcx_excel=>raise_text( lv_errormessage ).
          ENDTRY.
        ENDIF.
```  
I use lv_value to pass this value to methods SET_CELL.

Finally, to get these EDIT_MASK for each field automatically I've made the following changes:

Added this field in BIND_TABLE settings sctructure:
![image](https://github.com/abap2xlsx/abap2xlsx/assets/168567887/64e69475-6ba0-4761-a216-312aa5611a5d)

In method BIND_TABLE i've changed this line passing a new parameter to method GET_FIELDCATALOG

```
lt_field_catalog = zcl_excel_common=>get_fieldcatalog( ip_table = ip_table iv_get_edit_mask = ls_settings-get_edit_mask ).
```
And in method GET_FIELDCATALOG of class ZCL_EXCEL_COMMON I've created an optional parameter IV_GET_EDIT_MASK and add these lines:

    IF iv_get_edit_mask EQ abap_true. 
     <fcat>-edit_mask = ls_salv_t_column_ref-r_column->get_edit_mask( ).
    ENDIF.

With this you can obtain the edit_mask of each field automatically.
You can call this method GET_FIELDCATALOG before the method BIND_TABLE to obtain the fieldcat and then make some changes and pass the modified fieldcat to method BIND_TABLE, or you can call method BIND_TABLE without any fieldcat (inside this method GET_FIELDCATALOG is called and you can get the edit_mask or no depending on the settings).

Do you like some of these changes? Is it possible to do these things in other way and these changes are not necessary? 
Regards.

**Comment by sandraros** [comment]
Thank you. Difficult to say right now, but why not. The conversion exits is the biggest problem in abap2xlsx. In your discussion, you are quite precise, so you can directly create two issues (enhancements). There should be two pull requests. I'd prefer to discuss more after each PR is created, as I like to test and debug.

Summary of the two suggested enhancements:
1. Pass column width to `BIND_TABLE` (instead of subsequently calling repeatedly `SET_COLUMN_WIDTH`)
   - NB: to keep the semantic used by abap2xlsx consistent, call the new field "width" instead of "length".
2. Pass edit mask to `BIND_TABLE`.
   - If you only target the conversion exits, I'd prefer to limit the feature to them.

**Reply by RixarSAP** [reply]
Ok, I'm going to try to get access to a system with a newer version of Abap than mine. To be able to export the changes more easily. When I get it I will create the PR.

**Reply by gregorwolf** [reply]
When you contact me via the abapGit Slack (Invite at https://communityinviter.com/apps/abapgit/abap) I'm able to help you with access to an up-to-date system.

**Reply by RixarSAP** [reply]
I got access to a system with a more modern version of abap. 
But, without SSL, so I need to work offline, with the ZIP files...

I've downloaded and installed ZABAPGIT. and also installed ABAP2XLSX.
I've made some changes in the project and export them from transaction ZABAPGIT to a ZIP file. 

Now, how can I import this ZIP file into my github fork ?
I'm not using any git client, just the web browser.
https://docs.abapgit.org/user-guide/projects/offline/export-zip.html

I'm going to join the slack channel, if you don't mind, to ask for help. 
Thanks !

**Reply by gregorwolf** [reply]
Please follow https://docs.abapgit.org/user-guide/setup/ssl-setup.html or forward that to your basis admin to get SSL working. Otherwise its just pain.

**Reply by RixarSAP** [reply]
oook, I'll try.

**Reply by RixarSAP** [reply]
No answer from the BASIS team yet...  
I will create some simple PR modifying the files directly.
If I had access to a system with SSL it would be great to be able to contribute more easily.

**Reply by sandraros** [reply]
You may also:
1. Install Git on your laptop
2. Clone your fork on your laptop (folders and files will be replicated)
3. Modify the ABAP code on your ABAP server, download it as ZIP via abapGit on your laptop
4. Unzip on your laptop to replace the cloned files
5. Stage/Push from your laptop files to your fork
6. Create a Pull Request

**Reply by RixarSAP** [reply]
Thanks, Sandra.
I'll do this way. I've already created a new PR.

**Reply by RixarSAP** [reply]
1st PR for column WIDTH created and merged !

2nd is not necessary. Abap2XLSX already has a funcionality to perform conversions automatically. With parameter IP_CONV_EXIT_LENGHT of method BIND_TABLE is printing the values correctly doing the conversions...

I was trying in my old 701 Abap version system, that has a no an up to date version of ABAP2XLSX... and does not have this parameter...
